erc20-migration/contracts/ZTESTZendBackupVault.sol

```diff
@@ -0,0 +1,127 @@
+///         In the constructor will receive an admin address, the only entity authorized to perform load operations, and a cumulative hash 
+///         calcolated with all the dump data.
+contract ZTESTZendBackupVault  {
+
+    struct Balances {
+        uint256 amount;
+        bool distributed;
+    }
+    
+    // Map of the claimable balances.
+    // The key is the zendAddress in bs58 decoded format
+    mapping(bytes20 => Balances) public balances;
+       
+    // Cumulative Hash calculated
+    bytes32 private _cumulativeHash;
+
+    // Final Cumulative Hash filled in the constructor, used for checkpoint, to unlock claim
+    bytes32 private cumulativeHashCheckpoint;
+       
+    // admin authority
+    address public admin;
+
+    IERC20Mintable public zenToken;
+
+    string private constant MESSAGE_PREFIX = "ZENCLAIM";
+
+    error AddressNotValid();
+    error CumulativeHashNotValid();
+    error UnauthorizedOperation();
+    error ArrayLengthMismatch();
+    error ERC20NotSet();
+    error NothingToClaim(bytes20 zenAddress);
+    error AlreadyClaimed(bytes20 zenAddress);
+    event Claimed(address destAddress, bytes20 zenAddress, uint256 amount);
+
+
+    /// @notice Smart contract constructor
+    /// @param _admin  the only entity authorized to perform restore operations
+    /// @param _cumulativeHashCheckpoint  a cumulative recursive  hash calcolated with all the dump data.
+    ///                                   Will be used to verify the consinstency of the restored data, and as
+    ///                                   a checkpoint to understand when all the data has been loaded and the claim 
+    ///                                   can start
+    constructor(address _admin, bytes32 _cumulativeHashCheckpoint) {
+        if(_admin == address(0)) revert AddressNotValid();
+        if(_cumulativeHashCheckpoint == bytes32(0)) revert CumulativeHashNotValid();   
+        admin = _admin;
+        _cumulativeHash = bytes32(0);
+        cumulativeHashCheckpoint = _cumulativeHashCheckpoint;
+    }
+
+    /// @notice Insert a new bach of tuples (bytes20, value) and updates the cumulative hash.
+    ///         The zendAddresses in bs58 decoded format
+    ///         To guarantee the same algorithm is applied, the expected cumulativeHash after the batch processing must be provided explicitly)
+    function batchInsert(bytes32 expectedCumulativeHash, bytes20[] memory zendAddresses, uint256[] memory values) public {
+        if (msg.sender != admin) revert UnauthorizedOperation();               
+        if (zendAddresses.length != values.length) revert ArrayLengthMismatch();
+        uint256 i;
+        while (i != zendAddresses.length) {
+            balances[zendAddresses[i]] = Balances({amount: values[i], distributed: false});
+            _cumulativeHash = keccak256(abi.encode(_cumulativeHash, zendAddresses[i], values[i]));
+            unchecked { ++i; }
+        }
```

Comment:
```suggestion
        bytes32 auxHash = _cumulativeHash;
        while (i != zendAddresses.length) {
            balances[zendAddresses[i]] = Balances({amount: values[i], distributed: false});
            auxHash = keccak256(abi.encode(auxHash, zendAddresses[i], values[i]));
            unchecked { ++i; }
        }
        _cumulativeHash = auxHash;
```
A possible optimization, which avoids a costly storage write for each processed address. This is a minor optimization, but considering that the number of processed addresses will be very high (hundreds of thousands?), it could translate into non-negligible overall gas savings.